Adding all base files for meta-smartrack

HatsyRei opened the pull request:
Features include:
- Bitbake default build configuration
 - RCU base service recipe
 - RCU hostname service recipe and files
 - vsftpd recipe and configuration
 - Custom defconfig file
 - smartrack-minimal-image recipe

Practically all the same files in meta-smartrack-old.

HatsyRei committed:
Adding all base files for meta-smartrack. Features include
 - Bitbake default build configuration
 - RCU base service recipe
 - RCU hostname service recipe and files
 - vsftpd recipe and configuration
 - Custom defconfig file
 - smartrack-minimal-image recipe

karhin1990 commented on buildconf/export:
```diff
@@ -0,0 +1,46 @@
+echo "    tdx-reference-minimal-image"
+echo "    tdx-reference-multimedia-image"
+echo "    other (unsupported) targets may be found in"
+echo "    meta-toradex-demos/recipes-images/images/"
```
We may want to hide it as we don't want user to build toradex image and deploy it into Smart Racks RCU

karhin1990 commented on buildconf/export:
```diff
@@ -0,0 +1,46 @@
+			echo "# always get the newest kernel. Without use-head-next your build may fail."
+			echo "MACHINEOVERRIDES =. \"use-head-next:\""
+			echo ""
+		} > conf/local.conf.new
```
Since we don't support building "integration", is this piece of code still relevant?

karhin1990 commented on buildconf/export:
```diff
@@ -0,0 +1,46 @@
+	$ECHO -e "\033[1mA sample conf/local.conf file has been created"
+	$ECHO -e "Check and edit the file to adapt to your local needs\033[0m"
+	echo "The following likely need your attention:"
+	echo "DL_DIR"
```
We may need a clearer description here, or directly omit it if we don't want our client to worry about this.

We should also have comments that guide our client how to build an image, e.g. "To build an image, execute: 'bitbake <supported_image_target>'

karhin1990 commented on buildconf/local.conf:
```diff
@@ -0,0 +1,269 @@
+#MACHINE ?= "colibri-imx7-emmc"
+#MACHINE ?= "colibri-imx8x"
+#
+#MACHINE ?= "verdin-imx8mm"
```
We should remove other Toradex machine except apalis-imx8x because we don't want customer try to build those machine and run on our board,

karhin1990 commented on recipes-kernel/linux/files/defconfig:
```diff
@@ -0,0 +1,882 @@
+CONFIG_SYSVIPC=y
```
Can you briefly explain how this file get generated and what changes have you made on top of the toradex kernel config?

Also, should we name it as smart_racks_defconfig?

HatsyRei replied:
So far we haven't made any changes to the default kernel config yet (apart from an extra comment at the end of the file, to identify that this defconfig was used instead of the default one). We can change it to a different name, but this would most likely require some additional changes to be specified in our bbappend recipe as well. The KBUILD_DEFCONFIG_mx8 attribute in [Toradex's kernel recipe](https://git.toradex.com/cgit/meta-toradex-nxp.git/tree/recipes-kernel/linux/linux-toradex_5.4-2.1.x.bb?h=dunfell-5.x.y) points to 'defconfig' by default, so we'll have to try to change that.

karhin1990 commented on recipes-core/rcu-hostname/files/rcuhostname:
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+
+rcuserialpath=/usr/bin/rcu-serial
```
Can you put a TODO comment here indicates that this script/utility hasn't ready at the moment?

karhin1990 commented on recipes-core/rcu-hostname/files/rcuhostname:
```diff
@@ -0,0 +1,27 @@
+    serial="EMPTYSERIAL"
+fi
+
+hostname="NI-RCU-"${serial}
```
I would name it as NI-SmartRack-xxx because customer may not know what is actually an "RCU"

tanchunhau replied:
Seems like can just handled by **if , elseIf , else** (according to priority - RCU serial, SoM serial , EmptySerial):
Please refer to below :

```
#set serial to RCU serial number
if [ -f ${rcuserialpath} ]; then
    serial=$(tr -d '\0' <${rcuserialpath})

#set serial to Apalis SOM serial number
elif [ -f ${tdxserialpath} ]; then
     serial=$(tr -d '\0' <${tdxserialpath})

#set serial to "EMPTYSERIAL"
else
    serial="EMPTYSERIAL"
fi
```

HatsyRei replied:
Referring to Toradex's template for this script, it seems like there's a chance our serial file will exist but somehow empty or corrupted. If we proceed with the if elif else construct, I'm afraid we might not be able to account for that case.

karhin1990 commented on recipes-core/rcu-hostname/files/rcuhostname:
```diff
@@ -0,0 +1,27 @@
+#!/bin/sh
+
```
Can you put a comment on top of this file to mention the usage/purpose of this file?

HatsyRei committed:
Remove inapplicable machine options

HatsyRei committed:
Remove Toradex targets, integration build scripts

HatsyRei committed:
Changed hostname prefix, add descriptions, todo

tanchunhau reviewed (commented):
Note: 3x3 bit flags for a mode:
https://stackoverflow.com/questions/18415904/what-does-mode-t-0644-mean/18415935

HatsyRei committed:
Set bbfile priority from 6 to 24

Priority level 24 would put us on a similar level to Toradex image recipes stored in meta-toradex-demos

HatsyRei committed:
Add coreutils package to image recipe

Coreutils added mainly for test and utility purposes (e.g. dd). No impact to RCU services foreseen.

karhin1990 commented on conf/layer.conf:
```diff
@@ -0,0 +1,13 @@
+
+BBFILE_COLLECTIONS += "meta-smartrack"
+BBFILE_PATTERN_meta-smartrack = "^${LAYERDIR}/"
+BBFILE_PRIORITY_meta-smartrack = "24"
```
Can you place a brief comment here on how to come out this magic number (24)? It will help for future references and maintenance.

HatsyRei replied:
Added a comment above the line. I also changed priority from 24 to 25. My original intention was to allow meta-smartrack have priority over other Toradex layers. Layer priority 24 (assumed by meta-toradex-demos) was the highest among Toradex layers, so to achieve the aim I went one value higher.

karhin1990 commented on recipes-core/rcu-hostname/files/rcuhostname:
```diff
@@ -0,0 +1,30 @@
+#!/bin/sh
+
+# This script is responsible for reading the RCU serial and setting the RCU hostname accordingly. Paired with avahi-daemon which enables mDNS, this provides an option for clients to communicate with the RCU just by knowing its serial (instead of IP address).
```
... the RCU Serial **Number** and setting ...

karhin1990 commented on recipes-core/rcu-hostname/files/rcuhostname:
```diff
@@ -0,0 +1,30 @@
+    serial="EMPTYSERIAL"
+fi
+
+hostname="NI-SmartRack-"${serial}
```
Refer to Jonathan's comment on the feature spec, we need a way for user to set their custom hostname. 
I believe user will set the hostname via RCU Service, perhaps you need some placeholder code to read from a file (where RCU Service will write the user-defined hostname). If this file exist and the content is valid, we will use it as the hostname instead.

HatsyRei replied:
@karhin1990 I added a todo in this file, as well as a new task under the RCU image user story in AzDO.

karhin1990 reviewed (changes requested):
Overall looks solid, just minor things to address then we are good to go.

HatsyRei committed:
Alter layer priority

The highest layer priority specified among Toradex layers is 24:
https://git.toradex.com/cgit/meta-toradex-demos.git/tree/conf/layer.conf?h=dunfell-5.x.y

In order for some of our changes to take place, layer priority higher than 24 is required.

HatsyRei committed:
Minor changes to script description

HatsyRei committed:
Update TODO: Client-side hostname configuration
